Chips improvements

lyubomir-popov opened the pull request:
## Done

Based on #3175; improve styling of chips. Includes code for the popover menu. This is just to ensure the chips fit as expected in the bigger picture. This will be deleted before merging.

## QA

- Pull code
- Run `./run`
- Open docs/examples/patterns/chip/multiple
To evaluate in situ, add the following [scss](https://pastebin.canonical.com/p/Y7HqmMBFMD/) and replace the markup of any chips example with [this](https://pastebin.canonical.com/p/KZtFNSd8zW/)

- Verify it looks like this:

![image](https://user-images.githubusercontent.com/2741678/88532710-f3f74c00-cffc-11ea-97fc-ef745bdf167a.png)

Fixes #3179 
Fixes #3192

webteam-app commented:
Demo starting at https://vanilla-framework-3179.demos.haus

lyubomir-popov commented:
@anthonydillon I made some changes, mainly opted for display:inline on the chip and everything in it. 
The height of the chips wrapper  uses the height of individual chips and the margin between them to ensure precise fit.
I've also adjusted margins and paddings here and there to get it closer to the mockup.

lyubomir-popov commented:
@anthonydillon this is ready for another look

anthonydillon commented:
Unfortunately, display inline causes some wrapping issues on small screens:
![image](https://user-images.githubusercontent.com/1413534/88437173-cdb78d80-cdfd-11ea-9515-893a7ac09982.png)

anthonydillon commented:
The dismiss button has margin-bottom:
![Screenshot from 2020-07-24 22-34-55](https://user-images.githubusercontent.com/1413534/88437278-0ce5de80-cdfe-11ea-8e1c-6d1c41d4f9a7.png)

lyubomir-popov commented:
@anthonydillon updated to fix the inline issues, ready for another look.

anthonydillon commented on scss/_patterns_chip.scss:
```diff
@@ -1,18 +1,19 @@
+    display: inline-block;
+    line-height: $chip-line-height;
+    margin: $chip-margin-top $sph-inner--small 0 0;
     padding: $chip-vertical-padding $sph-inner;
```
I would suggest changing this from a result of:
```css
padding: 0 1rem;
```
to
```scss
padding-left: $sph-inner;
padding-right: $sph-inner;
```
Long-form is generally preferred over shorthand as its more specific and results it less complicated overrides.

anthonydillon commented on scss/_patterns_chip.scss:
```diff
@@ -1,18 +1,19 @@
-  $chip-height: map-get($line-heights, small) + 2 * $chip-vertical-padding;
-  $chip-margin-bottom: $spv-outer--small-scaleable - 2 * $chip-vertical-padding;
+  $chip-line-height: map-get($line-heights, default-text);
+  $chip-vertical-padding: 0;
```
All uses of this variable are obsolete so you can remove it once addressed below.

anthonydillon commented on scss/_patterns_chip.scss:
```diff
@@ -0,0 +1,75 @@
+    margin: $chip-margin-top $sph-inner--small 0 0;
+    padding: $chip-vertical-padding $sph-inner;
+    user-select: none;
+    white-space: nowrap;
```
It might also be nice to add overflow control with the following:
```css
max-width: 100%;
overflow: hidden;
text-overflow: ellipsis;
```

anthonydillon commented on scss/_patterns_chip.scss:
```diff
@@ -0,0 +1,75 @@
+
+      display: inline;
+      line-height: $chip-line-height;
+      position: relative;
```
I don't think you need display: relative here as your not changing the position.

lyubomir-popov commented:
@anthonydillon I've addressed the feedback above; with regards to overflow, I'm not sure we can properly solve this with the current markup:

![image](https://user-images.githubusercontent.com/2741678/88679612-807e3900-d0e7-11ea-831b-6f544d1ed547.png)

I think the two pieces of text (__lead and __value) need to be in a wrapper so I can fix the width of the wrapper to that of the chip - the width of the button. Otherwise, the dismiss button is pushed out of view and clipped by the overflow: hidden.

lyubomir-popov commented:
@anthonydillon Ok, created an issue to track the [overflow problem](https://github.com/canonical-web-and-design/vanilla-framework/issues/3192)

lyubomir-popov commented:
Hmm, overflow still needs some work in firefox:
![Uploading image.png…]()

lyubomir-popov commented:
@anthonydillon I think I finally got it to work:

![image](https://user-images.githubusercontent.com/2741678/88789772-0c976b80-d18f-11ea-8138-6863f8121fe7.png)

anthonydillon commented on templates/docs/examples/patterns/chip/multiple.html:
```diff
@@ -0,0 +1,12 @@
+
+{% block content %}
+
+<div class="p-chip">AWS</div>
```
Shouldn't these have `p-chip__value` wrapping them otherwise users wont use them.

lyubomir-popov commented:
@anthonydillon updated.

anthonydillon commented:
Thanks 👍
